# House Sales Prices Analysis Using Multi-Linear Regression

## <small>1. Businness Understanding</small>

### <small>a) Introduction</small>

The goal of this project is to analyze house sales data using multiple regression modelling techniques. By employing multiple regression, we aim to identify and quantify the relationships that exists between various factors or predictors and house sale prices.This analysis will provide insights into the key drivers of house prices and help shareholders make informed decisions related to real estate investment.

### <small> b) Problem Statement</small>

This project seek to address the importannce of identifying the factors that significantly affect house prices.Traditional approaches often rely on anectodal evidence or limited analysis, resulting into inaccurate insights. This knowledge gap hampers stakeholders' ability to predict and interpret house proce flactuations effectively. To address this problem, we propose utilizing multiple regression modeling techniques to analyze house sales data comprehensively. By conducting multiple regression, we can identify the relationship between different factors and house sales prices, taking into account the collective impact of multiple independent variables.

### <small>c) Main Objective</small>

To analyze house sales data using multiple regression modeling techniques to identify and quantify the key factors influencing house sales prices and help shareholders make data driven decisions on investing in real estate.

### <small>d) Subjective Objectives</small>

1. Conduct explanatory data analysis to gain insights on the relationships between different variables and target variable, assisting in selection of relevant variables for regression model.
2. Develop multiple regression model to predict house sale prices, considering the selected independent variables and their impact on the dependent variable.Validate the model assumptions, assess its goodness of fit and refine the model if necesssay.
3. Interpret the coefficients of the independent variables in the model to determine their individual impact on house prices, identifying the most influential factors driving thr house sales prices and their respective effects.
4. Evaluate and validate the performnance of the model.
5. Provide actionable insights and recommendations based on the analysis to assist real estate investors, and policymakers in making informed decisions regarding property investment, market trends, and economic planning.

### <small> e) Notebook Structure</small>

1. Reading the data.
2. Data Cleaning and Preprocessing.
3. Multiple Regression Modelling.
4. Model Evaluation and Undesrtanding.
5. Results Presentation and Conclusions.
6. Recommendations
7. References.

### <small> f) Data Understanding</small>

The dataset used in this project contains information about the factors affecting the housing prices including variables such as date, sqft_above, view and sqft_basement. Explanatory data analysis will be used to get a clear understanding of the dataset including the missing values, checking the data types, identifying outliers and also extracting relevant features for analysis. 

## 2. Importing Inportant Packages

In [353]:
import numpy as np # for algebra
import pandas as pd # for data processing
import warnings
warnings.filterwarnings('ignore')
import matplotlib.pyplot as plt
import seaborn as sns
from collections import Counter
from matplotlib import rcParams
rcParams['figure.figsize'] = 7, 5
pd.options.display.max_columns = 25
sns.set(style='darkgrid')

#validation
from sklearn.metrics import r2_score
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV,RandomizedSearchCV
#models 
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# Any results you write to the current directory are saved as output.rom sklearn.preprocessing import StandardScaler



## 3. Reading Data

In [354]:
df = pd.read_csv('data/kc_house_data.csv')
# See the head of the dataframe
df.head()

id        date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  10/13/2014  221900.0         3       1.00         1180   
1  6414100192   12/9/2014  538000.0         3       2.25         2570   
2  5631500400   2/25/2015  180000.0         2       1.00          770   
3  2487200875   12/9/2014  604000.0         4       3.00         1960   
4  1954400510   2/18/2015  510000.0         3       2.00         1680   

   sqft_lot  floors waterfront  view  condition          grade  sqft_above  \
0      5650     1.0        NaN  NONE    Average      7 Average        1180   
1      7242     2.0         NO  NONE    Average      7 Average        2170   
2     10000     1.0         NO  NONE    Average  6 Low Average         770   
3      5000     1.0         NO  NONE  Very Good      7 Average        1050   
4      8080     1.0         NO  NONE    Average         8 Good        1680   

  sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
0           0.0      1955           0.0    98178  47.5112 -122.257   
1         400.0      1951        1991.0    98125  47.7210 -122.319   
2           0.0      1933           NaN    98028  47.7379 -122.233   
3         910.0      1965           0.0    98136  47.5208 -122.393   
4           0.0      1987           0.0    98074  47.6168 -122.045   

   sqft_living15  sqft_lot15  
0           1340        5650  
1           1690        7639  
2           2720        8062  
3           1360        5000  
4           1800        7503

In [355]:
df['floors'].value_counts()

1.0    10673
2.0     8235
1.5     1910
3.0      611
2.5      161
3.5        7
Name: floors, dtype: int64

## 4. Explanatory Data Analysis

In [356]:
# Check the shape of the dataset
print(f"Shape of df is : {df.shape}")

Shape of df is : (21597, 21)


In [357]:
# An overview of the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21597 entries, 0 to 21596
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21597 non-null  int64  
 1   date           21597 non-null  object 
 2   price          21597 non-null  float64
 3   bedrooms       21597 non-null  int64  
 4   bathrooms      21597 non-null  float64
 5   sqft_living    21597 non-null  int64  
 6   sqft_lot       21597 non-null  int64  
 7   floors         21597 non-null  float64
 8   waterfront     19221 non-null  object 
 9   view           21534 non-null  object 
 10  condition      21597 non-null  object 
 11  grade          21597 non-null  object 
 12  sqft_above     21597 non-null  int64  
 13  sqft_basement  21597 non-null  object 
 14  yr_built       21597 non-null  int64  
 15  yr_renovated   17755 non-null  float64
 16  zipcode        21597 non-null  int64  
 17  lat            21597 non-null  float64
 18  long  

In [358]:
# Statistical distribjtion of the dataset
df.describe()

id         price      bedrooms     bathrooms   sqft_living  \
count  2.159700e+04  2.159700e+04  21597.000000  21597.000000  21597.000000   
mean   4.580474e+09  5.402966e+05      3.373200      2.115826   2080.321850   
std    2.876736e+09  3.673681e+05      0.926299      0.768984    918.106125   
min    1.000102e+06  7.800000e+04      1.000000      0.500000    370.000000   
25%    2.123049e+09  3.220000e+05      3.000000      1.750000   1430.000000   
50%    3.904930e+09  4.500000e+05      3.000000      2.250000   1910.000000   
75%    7.308900e+09  6.450000e+05      4.000000      2.500000   2550.000000   
max    9.900000e+09  7.700000e+06     33.000000      8.000000  13540.000000   

           sqft_lot        floors    sqft_above      yr_built  yr_renovated  \
count  2.159700e+04  21597.000000  21597.000000  21597.000000  17755.000000   
mean   1.509941e+04      1.494096   1788.596842   1970.999676     83.636778   
std    4.141264e+04      0.539683    827.759761     29.375234    399.946414   
min    5.200000e+02      1.000000    370.000000   1900.000000      0.000000   
25%    5.040000e+03      1.000000   1190.000000   1951.000000      0.000000   
50%    7.618000e+03      1.500000   1560.000000   1975.000000      0.000000   
75%    1.068500e+04      2.000000   2210.000000   1997.000000      0.000000   
max    1.651359e+06      3.500000   9410.000000   2015.000000   2015.000000   

            zipcode           lat          long  sqft_living15     sqft_lot15  
count  21597.000000  21597.000000  21597.000000   21597.000000   21597.000000  
mean   98077.951845     47.560093   -122.213982    1986.620318   12758.283512  
std       53.513072      0.138552      0.140724     685.230472   27274.441950  
min    98001.000000     47.155900   -122.519000     399.000000     651.000000  
25%    98033.000000     47.471100   -122.328000    1490.000000    5100.000000  
50%    98065.000000     47.571800   -122.231000    1840.000000    7620.000000  
75%    98118.000000     47.678000   -122.125000    2360.000000   10083.000000  
max    98199.000000     47.777600   -121.315000    6210.000000  871200.000000

## 5. Data Wrangling

### <small> a) Data Cleaning </small>

In [359]:
# Check for null values
pd.DataFrame(df.isna().sum()).T


id  date  price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0   0     0      0         0          0            0         0       0   

   waterfront  view  condition  grade  sqft_above  sqft_basement  yr_built  \
0        2376    63          0      0           0              0         0   

   yr_renovated  zipcode  lat  long  sqft_living15  sqft_lot15  
0          3842        0    0     0              0           0

In [360]:
df['waterfront'].value_counts()

NO     19075
YES      146
Name: waterfront, dtype: int64

In [361]:
# Fill all the null values with 'Unknown' since the column contains categorical data
df['waterfront'].fillna('NO',inplace=True)
# Check for missing values now
pd.DataFrame(df.isna().sum()).T

id  date  price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0   0     0      0         0          0            0         0       0   

   waterfront  view  condition  grade  sqft_above  sqft_basement  yr_built  \
0           0    63          0      0           0              0         0   

   yr_renovated  zipcode  lat  long  sqft_living15  sqft_lot15  
0          3842        0    0     0              0           0

In [362]:
df['waterfront'].unique()

array(['NO', 'YES'], dtype=object)

In [363]:
# Deal with missing values in the view column
# Check for the counts of different categories
df['view'].value_counts()


NONE         19422
AVERAGE        957
GOOD           508
FAIR           330
EXCELLENT      317
Name: view, dtype: int64

In [364]:
#  Replace the missing values with the most frequent category
df['view'].fillna('NONE',inplace=True)

In [365]:
# check forthe count unique values in the year renoveted column 
df['yr_renovated'].value_counts()

0.0       17011
2014.0       73
2013.0       31
2003.0       31
2007.0       30
          ...  
1951.0        1
1953.0        1
1946.0        1
1976.0        1
1948.0        1
Name: yr_renovated, Length: 70, dtype: int64

In [366]:
# Replace null with the most frequent value
df['yr_renovated'].fillna(0,inplace=True)

In [367]:
# Check for null values
pd.DataFrame(df.isna().sum()).T

id  date  price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0   0     0      0         0          0            0         0       0   

   waterfront  view  condition  grade  sqft_above  sqft_basement  yr_built  \
0           0     0          0      0           0              0         0   

   yr_renovated  zipcode  lat  long  sqft_living15  sqft_lot15  
0             0        0    0     0              0           0

No missing values!

In [368]:
#iterate through the columns to see the frequency of different values
for i in df.columns:
    print(df[i].value_counts())

795000620     3
8910500150    2
7409700215    2
1995200200    2
9211500620    2
             ..
3649100387    1
2767603649    1
1446403617    1
5602000275    1
1523300157    1
Name: id, Length: 21420, dtype: int64
6/23/2014    142
6/25/2014    131
6/26/2014    131
7/8/2014     127
4/27/2015    126
            ... 
11/2/2014      1
1/31/2015      1
5/24/2015      1
5/17/2014      1
7/27/2014      1
Name: date, Length: 372, dtype: int64
450000.0    172
350000.0    172
550000.0    159
500000.0    152
425000.0    150
           ... 
278800.0      1
439888.0      1
354901.0      1
942000.0      1
402101.0      1
Name: price, Length: 3622, dtype: int64
3     9824
4     6882
2     2760
5     1601
6      272
1      196
7       38
8       13
9        6
10       3
11       1
33       1
Name: bedrooms, dtype: int64
2.50    5377
1.00    3851
1.75    3048
2.25    2047
2.00    1930
1.50    1445
2.75    1185
3.00     753
3.50     731
3.25     589
3.75     155
4.00     136
4.50     100
4.25      79
0.

### b) Encoding Categorical Columns in the Data

In [369]:
# Define the order of the categories
category_order = {
    'NO': 0,
    "YES" : 1
}

# Perform ordinal encoding on the "waterfront" column
df["waterfront_encoded"] = df["waterfront"].map(category_order)

# Replace the original "waterfront" column with the encoded values
df["waterfront"] = df["waterfront_encoded"]

# Check the updated dataset
df['waterfront'].value_counts()



0    21451
1      146
Name: waterfront, dtype: int64

In [370]:
# Let's check the count of the encoded column of waterfront
df['waterfront'].value_counts()

0    21451
1      146
Name: waterfront, dtype: int64

In [371]:
# Encode the column view
# Define the order of the categories
category_order = {
    "NONE": 0,
    "AVERAGE": 1,
    "GOOD": 2,
    "FAIR": 3,
    "EXCELLENT": 4
}

# Perform ordinal encoding on the "view" column
df["view_encoded"] = df["view"].map(category_order)

# Replace the original "view" column with the encoded values
df["view"] = df["view_encoded"]

# Check the updated dataset
df['view'].value_counts()


0    19485
1      957
2      508
3      330
4      317
Name: view, dtype: int64

In [372]:
# Let's encode the condition column 
# Define the order of the categories
category_order = {
    "Poor": 1,
    "Fair": 2,
    "Average": 3,
    "Good": 4,
    "Very Good": 5
}

# Perform ordinal encoding on the "condition" column
df["condition_encoded"] = df["condition"].map(category_order)

# Replace the original "condition" column with the encoded values
df["condition"] = df["condition_encoded"]

# Check the updated dataset
df['condition'].value_counts()

3    14020
4     5677
5     1701
2      170
1       29
Name: condition, dtype: int64

In [373]:
# Encode the grade column
# Define the order of the categories
category_order = {
    "3 Poor": 3,
    "5 Fair": 5,
    "4 Low": 4,
    "6 Low Average": 6,
    "7 Average": 7,
    "8 Good": 8,
    "9 Better": 9,
    "10 Very Good": 10,
    "11 Excellent": 11,
    "12 Luxury": 12,
    "13 Mansion": 13
}

# Perform ordinal encoding on the "grade" column
df["grade_encoded"] = df["grade"].replace(category_order)

# Replace the original "grade" column with the encoded values
df["grade"] = df["grade_encoded"]

# Check the updated dataset
df['grade'].value_counts()

7     8974
8     6065
9     2615
6     2038
10    1134
11     399
5      242
12      89
4       27
13      13
3        1
Name: grade, dtype: int64

In [374]:
# View the encoded dataset
df_encoded = df
df_encoded

id        date     price  bedrooms  bathrooms  sqft_living  \
0      7129300520  10/13/2014  221900.0         3       1.00         1180   
1      6414100192   12/9/2014  538000.0         3       2.25         2570   
2      5631500400   2/25/2015  180000.0         2       1.00          770   
3      2487200875   12/9/2014  604000.0         4       3.00         1960   
4      1954400510   2/18/2015  510000.0         3       2.00         1680   
...           ...         ...       ...       ...        ...          ...   
21592   263000018   5/21/2014  360000.0         3       2.50         1530   
21593  6600060120   2/23/2015  400000.0         4       2.50         2310   
21594  1523300141   6/23/2014  402101.0         2       0.75         1020   
21595   291310100   1/16/2015  400000.0         3       2.50         1600   
21596  1523300157  10/15/2014  325000.0         2       0.75         1020   

       sqft_lot  floors  waterfront  view  condition  grade  sqft_above  \
0          5650     1.0           0     0          3      7        1180   
1          7242     2.0           0     0          3      7        2170   
2         10000     1.0           0     0          3      6         770   
3          5000     1.0           0     0          5      7        1050   
4          8080     1.0           0     0          3      8        1680   
...         ...     ...         ...   ...        ...    ...         ...   
21592      1131     3.0           0     0          3      8        1530   
21593      5813     2.0           0     0          3      8        2310   
21594      1350     2.0           0     0          3      7        1020   
21595      2388     2.0           0     0          3      8        1600   
21596      1076     2.0           0     0          3      7        1020   

      sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
0               0.0      1955           0.0    98178  47.5112 -122.257   
1             400.0      1951        1991.0    98125  47.7210 -122.319   
2               0.0      1933           0.0    98028  47.7379 -122.233   
3             910.0      1965           0.0    98136  47.5208 -122.393   
4               0.0      1987           0.0    98074  47.6168 -122.045   
...             ...       ...           ...      ...      ...      ...   
21592           0.0      2009           0.0    98103  47.6993 -122.346   
21593           0.0      2014           0.0    98146  47.5107 -122.362   
21594           0.0      2009           0.0    98144  47.5944 -122.299   
21595           0.0      2004           0.0    98027  47.5345 -122.069   
21596           0.0      2008           0.0    98144  47.5941 -122.299   

       sqft_living15  sqft_lot15  waterfront_encoded  view_encoded  \
0               1340        5650                   0             0   
1               1690        7639                   0             0   
2               2720        8062                   0             0   
3               1360        5000                   0             0   
4               1800        7503                   0             0   
...              ...         ...                 ...           ...   
21592           1530        1509                   0             0   
21593           1830        7200                   0             0   
21594           1020        2007                   0             0   
21595           1410        1287                   0             0   
21596           1020        1357                   0             0   

       condition_encoded  grade_encoded  
0                      3              7  
1                      3              7  
2                      3              6  
3                      5              7  
4                      3              8  
...                  ...            ...  
21592                  3              8  
21593                  3              8  
21594                  3              7  
21595                  3              8

In [376]:
# Convert "yr_renovated" column to integer
df_encoded["yr_renovated"] = df_encoded["yr_renovated"].astype(int)

In [378]:
# Drop the encoded columns
df_updated= df_encoded.drop(["waterfront_encoded", "view_encoded", "condition_encoded", "grade_encoded"], axis=1)

# Check the updated dataset
df_updated

id        date     price  bedrooms  bathrooms  sqft_living  \
0      7129300520  10/13/2014  221900.0         3       1.00         1180   
1      6414100192   12/9/2014  538000.0         3       2.25         2570   
2      5631500400   2/25/2015  180000.0         2       1.00          770   
3      2487200875   12/9/2014  604000.0         4       3.00         1960   
4      1954400510   2/18/2015  510000.0         3       2.00         1680   
...           ...         ...       ...       ...        ...          ...   
21592   263000018   5/21/2014  360000.0         3       2.50         1530   
21593  6600060120   2/23/2015  400000.0         4       2.50         2310   
21594  1523300141   6/23/2014  402101.0         2       0.75         1020   
21595   291310100   1/16/2015  400000.0         3       2.50         1600   
21596  1523300157  10/15/2014  325000.0         2       0.75         1020   

       sqft_lot  floors  waterfront  view  condition  grade  sqft_above  \
0          5650     1.0           0     0          3      7        1180   
1          7242     2.0           0     0          3      7        2170   
2         10000     1.0           0     0          3      6         770   
3          5000     1.0           0     0          5      7        1050   
4          8080     1.0           0     0          3      8        1680   
...         ...     ...         ...   ...        ...    ...         ...   
21592      1131     3.0           0     0          3      8        1530   
21593      5813     2.0           0     0          3      8        2310   
21594      1350     2.0           0     0          3      7        1020   
21595      2388     2.0           0     0          3      8        1600   
21596      1076     2.0           0     0          3      7        1020   

      sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
0               0.0      1955             0    98178  47.5112 -122.257   
1             400.0      1951          1991    98125  47.7210 -122.319   
2               0.0      1933             0    98028  47.7379 -122.233   
3             910.0      1965             0    98136  47.5208 -122.393   
4               0.0      1987             0    98074  47.6168 -122.045   
...             ...       ...           ...      ...      ...      ...   
21592           0.0      2009             0    98103  47.6993 -122.346   
21593           0.0      2014             0    98146  47.5107 -122.362   
21594           0.0      2009             0    98144  47.5944 -122.299   
21595           0.0      2004             0    98027  47.5345 -122.069   
21596           0.0      2008             0    98144  47.5941 -122.299   

       sqft_living15  sqft_lot15  
0               1340        5650  
1               1690        7639  
2               2720        8062  
3               1360        5000  
4               1800        7503  
...              ...         ...  
21592           1530        1509  
21593           1830        7200  
21594           1020        2007  
21595           1410        1287  
21596           1020        1357  

[21597 rows x 21 columns]

In [379]:
# Check the correlation between the independent factors and the target variable house prices
pd.DataFrame(df_updated.corr()['price']).sort_values(by='price',ascending=False)

price
price          1.000000
sqft_living    0.701917
grade          0.667951
sqft_above     0.605368
sqft_living15  0.585241
bathrooms      0.525906
view           0.386430
bedrooms       0.308787
lat            0.306692
waterfront     0.264306
floors         0.256804
yr_renovated   0.117855
sqft_lot       0.089876
sqft_lot15     0.082845
yr_built       0.053953
condition      0.036056
long           0.022036
id            -0.016772
zipcode       -0.053402

From the results above sqft_living has the highest positive correlation coefficient (0.701917), indicating a relatively strong positive linear relationship with the price of a property. As the square footage of the living area increases, the price tends to increase as well.
sqft_above also has a positive correlation coefficient of 0.605368, indicating a moderately strong positive linear relationship with the price. It represents the square footage of the interior living space above ground level.
sqft_lot has a weak positive correlation coefficient of 0.089876. It represents the square footage of the land area of the property.
id: This feature has a weak negative correlation coefficient of -0.016772. It represents the identifier of the property, which is not meaningful in terms of correlation with price.
zipcode has a weak negative correlation coefficient of -0.053402. It represents the zipcode of the property, which also does not show a strong correlation with price.

Keep in mind that correlation does not imply causation. These correlation coefficients only indicate the linear relationship between the features and the price in the dataset, and there may be other factors not captured in the dataset that affect the price of a property